pkg/autodiscovery/providers/kubecontainer.go

```diff
@@ -58,212 +49,158 @@ func (k *KubeContainerConfigProvider) String() string {
+			}
 
-func (k *KubeContainerConfigProvider) addEntity(entity workloadmeta.Entity) {
-	switch e := entity.(type) {
-	case *workloadmeta.KubernetesPod:
-		k.addPod(e)
-	case *workloadmeta.Container:
-		k.RLock()
-		_, containerSeen := k.containerCache[e.ID]
-		k.RUnlock()
-		if containerSeen {
-			// This is for short-lived detection, see `deleteEntity`.
-			//
-			// If we get a 'start' event for a container we've seen before,
-			// then it is likely (guaranteed?) that we saw a 'delete' event before this.
-			//
-			// The containerCache entry is still present which means that the
-			// deletion is currently "pending" and the actual deletion will occur in the
-			// the future, specifically in [0, delayDuration) amount of time.
-			// To guarantee our 'add' operation is processed _after_ the deletion occurs,
-			// set delay for 'delayDuration'
-			time.AfterFunc(delayDuration, func() {
-				k.addContainer(e)
-			})
-		} else {
-			k.addContainer(e)
-		}
+			oldConfigs, found := k.configCache[configID]
+			if found {
+				for digest, config := range oldConfigs {
+					_, ok := configsToAdd[digest]
+					if ok {
+						delete(configsToAdd, digest)
+					} else {
+						delete(k.configCache[configID], digest)
+						changes.Unschedule = append(changes.Unschedule, config)
+					}
+				}
+			} else {
+				k.configCache[configID] = configsToAdd
+			}
 
-	}
-}
+			for _, config := range configsToAdd {
+				changes.Schedule = append(changes.Schedule, config)
+			}
 
-func (k *KubeContainerConfigProvider) addPod(p *workloadmeta.KubernetesPod) {
-	k.Lock()
-	defer k.Unlock()
-	id := p.GetID().ID
-	k.podCache[id] = p
-	log.Debugf("Adding entity pod with ID %s\n", id)
-	k.upToDate = false
-}
+		case workloadmeta.EventTypeUnset:
+			oldConfigs, found := k.configCache[configID]
+			if !found {
+				log.Debugf("entity %q removed from workloadmeta store but not found in cache. skipping", configID)
+				continue
+			}
 
-func (k *KubeContainerConfigProvider) addContainer(c *workloadmeta.Container) {
-	k.Lock()
-	defer k.Unlock()
-	k.containerCache[c.ID] = c
-	log.Debugf("Adding entity container with ID %s\n", c.ID)
-	k.upToDate = false
-}
+			for _, oldConfig := range oldConfigs {
+				changes.Unschedule = append(changes.Unschedule, oldConfig)
+			}
 
-func (k *KubeContainerConfigProvider) deleteEntity(entity workloadmeta.Entity) {
-	switch e := entity.(type) {
-	case *workloadmeta.KubernetesPod:
-		k.deletePod(e)
-	case *workloadmeta.Container:
-		// Delay for short-lived detection
-		time.AfterFunc(delayDuration, func() {
-			k.deleteContainer(e)
-		})
-	}
-}
+			delete(k.configCache, configID)
+			delete(k.configErrors, configID)
 
-func (k *KubeContainerConfigProvider) deletePod(p *workloadmeta.KubernetesPod) {
-	k.Lock()
-	defer k.Unlock()
-	delete(k.podCache, p.GetID().ID)
-	log.Debugf("Deleting entity pod with ID %s\n", p.GetID().ID)
-	k.upToDate = false
-}
+		default:
+			log.Errorf("cannot handle event of type %d", event.Type)
+		}
```

Comment:
I'm wondering if `processEvents()` should return a `*integration.ConfigChanges`, and return `nil` if the default case.

Also send to outCh only if not nil here https://github.com/DataDog/datadog-agent/pull/12968/files#diff-df5e79509b4a5e5fc039e33b8e6a19c75666ee176b26f611e1f64045f81ada29R81

Reply:
Due to migrating `configChanges` to `integration.ConfigChanges`, now we have `IsEmpty`, so I'm using that to prevent sending an empty set of change to `outCh`.